### PR TITLE
redis: fix stream life time issues in key formatter

### PR DESCRIPTION
**Pull request by ramaraochavali**
Commit Message: fix stream life time issues in key formatter
Additional Description: Without this, if the downstream connection/stream closes before the key formatter kicks, Envoy crashes while getting the streamInfo from closed downstream connection.
Risk Level: Low
Testing: Updated. Test that crashes without the fix.
Docs Changes: N/A
Release Notes: Updated

### CODE REVIEW

File: test/extensions/filters/network/redis_proxy/router_impl_test.cc [hunk 1]
```diff
@@ -152,12 +153,60 @@ TEST(PrefixRoutesTest, TestFormatterWithPrefixRoute) {
...
   }
   PrefixRoutes router(prefix_routes, std::move(upstreams), runtime_);
-  router.setReadFilterCallback(&filter_callbacks);
+  router.initializeReadFilterCallbacks(&filter_callbacks);
   std::string key("abc:bar");
   EXPECT_EQ(upstream_c, router.upstreamPool(key)->upstream());
   EXPECT_EQ("{abc:bar}-test-subjectCN-{abc:bar}", key);
 }
 
+// This test validates the formatter works correctly when downstream is connection is closed
```
Comment (deveshkandpal1224) [inline test/extensions/filters/network/redis_proxy/router_impl_test.cc:162]: nit: `when downstream connection is closed`

**Comment by yanavlasov** [comment]
@ramaraochavali please fix merge conflict.

/wait

**Comment by ramaraochavali** [comment]
@yanavlasov done. PTAL

**Comment by yanavlasov** [comment]
I'm not very familiar with Redis. Pinging @mattklein123 and @weisisea as current owners.

File: source/extensions/filters/network/redis_proxy/router_impl.h [hunk 2]
```diff
@@ -64,16 +66,21 @@ class Prefix : public Route {
...
                Upstreams&& upstreams, Runtime::Loader& runtime);
 
   RouteSharedPtr upstreamPool(std::string& key) override;
 
-  void setReadFilterCallback(Network::ReadFilterCallbacks* callbacks) override {
-    callbacks_ = callbacks;
+  void initializeReadFilterCallbacks(Network::ReadFilterCallbacks* callbacks) override {
+    // Capture the stream info so that we can use it for formatting. This is needed to
+    // to avoid lifetime issues if the stream is destroyed before formatting is done.
+    stream_info_ = std::make_unique<StreamInfo::StreamInfoImpl>(
```
Comment (ramaraochavali) [inline source/extensions/filters/network/redis_proxy/router_impl.h:80]: We are facing few issues copying stream like this based on passed in callbacks. 

It seems the connectionProvider is getting mixed up with new connections. It happens before this change also (previously we were holding callbacks reference here).  Not exactly sure why though. May be holding callbacks reference here is not a good idea. We are experimenting by changing to pass callbacks in `makeRequest` and through the call chain to `upstreamPool` method.

Any other suggestions?

**Comment by ramaraochavali** [comment]
Found a better way to fix this. I will close this PR and open another PR

**Pull request closed without merging**